### PR TITLE
feat: Add Angle Protocol bridge

### PR DESCRIPTION
**Pull request by teddav**
# Description

Bridge for [Angle Protocol](https://angle.money/). Allows user to deposit tokens and receive sanTokens in exchange (yield bearing).

# Checklist:

- [x] I have reviewed my diff in github, line by line.
- [x] Every change is related to the PR description.
- [x] There are no unexpected formatting changes, or superfluous debug logs.
- [x] The branch has been rebased against the head of its merge target.
- [x] I'm happy for the PR to be merged at the reviewers next convenience.
- [x] NatSpec documentation of all the non-test functions is present and is complete.
- [x] Continuous integration (CI) passes.
- [ ] Command `forge coverage --match-contract MyContract` returns 100% line coverage.
- [x] All the possible reverts are tested.

### CODE REVIEW

File: src/bridges/angle/AngleBridge.sol
Comment (benesjan) [inline src/bridges/angle/AngleBridge.sol:92]: We can do a gas optimization here. If you would subtract dust amount (typically 1) from the balance then the storage slot would never reset to 0 which would make it more efficient to use. This is caused by [EIP-1087](EIP-1087). We use this optimization for example in the [Liquity's StakingBridge](https://github.com/AztecProtocol/aztec-connect-bridges/blob/6388b08be07ebd4d86a2de8e6bf2ca3c2cc8ac5b/src/bridges/liquity/StakingBridge.sol#L174).

**Review by benesjan — COMMENTED** [review]
The code looks good. But would be useful to expand the tests so that they also include "E2E" tests. How the E2E tests work is that we don't call the convert function directly but we instead call it through the RollupProcessor. Check out our [example test](https://github.com/AztecProtocol/aztec-connect-bridges/blob/6388b08be07ebd4d86a2de8e6bf2ca3c2cc8ac5b/src/test/bridges/example/ExampleE2E.t.sol#L53) for how to do those. E2E tests are important because it allows us to test whether bridge gas limit and approvals are set correctly.

File: src/bridges/angle/AngleBridge.sol
Comment (teddav) [reply src/bridges/angle/AngleBridge.sol]: thanks for the tip!

**Comment by teddav** [comment]
@benesjan thanks for the review! I just pushed a new commit with the changes.
Added E2E tests, and left some `dust` on the contract to avoid resetting `balanceOf` storage slots

File: src/test/bridges/angle/Angle.t.sol
Comment (benesjan) [inline src/test/bridges/angle/Angle.t.sol:250]: Here would be a good idea to use `bound(...)` method from _forge-std_ instead of `vm.assume(...)` because `vm.assume(...)` makes too many of the fuzz runs skip the test here (range is limited too much). We use bound a lot in other tests so feel free to check it out.

**Review by benesjan — CHANGES_REQUESTED** [review]
Other than using `bound(...)` instead of `vm.assume(...)` in that one test all looks good to me. Before we merge this it would be good to add tests of the angle-bridge-data.ts.

File: src/bridges/angle/AngleBridge.sol
Comment (benesjan) [inline src/bridges/angle/AngleBridge.sol:23]: Another gas optimization idea which came to my mind is replacing this mapping with `getPoolManager(...)` function which would simply contain few if statements with hardcoded addresses. That way we would avoid doing the SLOAD since those values would simply get copied from bytecode. But I leave that up to you if you want to do it. I did the same optimization in the [Uniswap bridge](https://github.com/AztecProtocol/aztec-connect-bridges/blob/80a8b3dbf1b2c5e302fc2613ec7588108269da34/src/bridges/uniswap/UniswapBridge.sol#L549) and it significantly decreased the costs.

File: src/test/bridges/angle/Angle.t.sol
Comment (benesjan) [inline src/test/bridges/angle/Angle.t.sol:40]: Casting the address to address everywhere in the tests is unnecessary but it's just a detail.

File: src/test/bridges/angle/Angle.t.sol
Comment (benesjan) [inline src/test/bridges/angle/Angle.t.sol:38]: Can you please set the gas limit as low as possible here? If the value is high then the users would overpay since they get charged the full gas limit by the rollup processor.

File: src/test/bridges/angle/Angle.t.sol
Comment (benesjan) [inline src/test/bridges/angle/Angle.t.sol:39]: The same applies to these assets. The value here represents how much gas gets consumed during token transfer.

File: src/bridges/angle/AngleBridge.sol
Comment (benesjan) [inline src/bridges/angle/AngleBridge.sol:77]: Subtracting the DUST here would mean that the dust would keep on accumulating in the bridge. I think in this case it would be better to not subtract the value here and instead I would send the value in a separate transaction to the bridge after deployment. In the [other tests](https://github.com/AztecProtocol/aztec-connect-bridges/blob/80a8b3dbf1b2c5e302fc2613ec7588108269da34/src/test/bridges/liquity/StakingBridgeUnit.t.sol#L26) we simulate it by `dealing` that value to the bridge in the setup function.

File: src/bridges/angle/AngleBridge.sol
Comment (benesjan) [inline src/bridges/angle/AngleBridge.sol:95]:
```suggestion
        } else {
            revert ErrorLib.InvalidAuxData();
        }
```
Here would be a good idea to revert if `_auxData` is not 0 or 1.

**Review by benesjan — CHANGES_REQUESTED** [review]
Just few small comments. The only bigger thing missing are the tests of angle-bridge-data.ts.

File: src/bridges/angle/AngleBridge.sol
Comment (benesjan) [reply src/bridges/angle/AngleBridge.sol:23]: Running the tests without this optimization resulted in max consumption of 138562 gas. Let's see how big a difference this will make.
image

**Review by benesjan — CHANGES_REQUESTED** [review]
All looks good to me know. Last thing is to add tests of the typescript class. Once you do that I will sent it my colleague for a review as well and then we'll merge it. Thank you

File: src/client/angle/angle-bridge-data.ts
Comment (LHerskind) [inline src/client/angle/angle-bridge-data.ts:15]: Why is there no function for APR etc? Seems like an important metric for a yield bearing asset.

File: src/test/bridges/angle/Angle.t.sol
Comment (LHerskind) [inline src/test/bridges/angle/Angle.t.sol:14]: Please consider making two separate test files for unit and E2E tests.

Also, tests are only done with Dai, but you have multiple collaterals for the bridge, test more than once, especially when you have a different number of decimals across the assets.

File: src/test/bridges/angle/Angle.t.sol
Comment (LHerskind) [inline src/test/bridges/angle/Angle.t.sol:100]: Please use the fuzzer here to not just have a fixed deposit amount that works.

File: src/test/bridges/angle/Angle.t.sol
Comment (LHerskind) [inline src/test/bridges/angle/Angle.t.sol:101]: You don't need to prank all of it, only the convert really needs to be pranked.

File: src/test/bridges/angle/Angle.t.sol
Comment (LHerskind) [inline src/test/bridges/angle/Angle.t.sol:122]: Use a constant instead of inserting the address directly.

File: src/test/bridges/angle/Angle.t.sol
Comment (LHerskind) [inline src/test/bridges/angle/Angle.t.sol:169]: Use the fuzzer for amount.

File: src/test/bridges/angle/Angle.t.sol
Comment (LHerskind) [inline src/test/bridges/angle/Angle.t.sol:170]: It is not required to prank the entire thing

File: src/test/bridges/angle/Angle.t.sol
Comment (LHerskind) [inline src/test/bridges/angle/Angle.t.sol:189]: Constant with name instead

**Review by LHerskind — CHANGES_REQUESTED** [review]

File: src/bridges/angle/AngleBridge.sol
Comment (teddav) [reply src/bridges/angle/AngleBridge.sol:72]: It looks "almost" similar but it's not. Every line is different, going back and forth between input and output asset. So i think it's better to leave it like this, otherwise I would have an `if` on every line (or a ternary) and it will look confusing

File: src/bridges/angle/AngleBridge.sol
Comment (teddav) [reply src/bridges/angle/AngleBridge.sol:108]: yes true. since we decided to hardcode everything we would need a different bridge to update

File: src/bridges/angle/AngleBridge.sol
Comment (teddav) [reply src/bridges/angle/AngleBridge.sol]: No, you're right this value will not change. I hardcoded it. It will be better

**Comment by teddav** [comment]
Thanks for the review @LHerskind . I made some changes based on your comments. Others, I think are not necessary. I still have some changes to do, and to write the typescript tests for the AngleBridgeData class

**Comment by teddav** [comment]
@LHerskind let me know when you have time to go over my changes/comments

**Comment by teddav** [comment]
@LHerskind just pushed some changes

File: src/bridges/angle/AngleSLPBridge.sol
Comment (LHerskind) [inline src/bridges/angle/AngleSLPBridge.sol:154]: Seems unnecessary to make an internal function of 2 lines that is used only once.

File: src/test/bridges/angle/AngleSLPE2E.t.sol
Comment (LHerskind) [inline src/test/bridges/angle/AngleSLPE2E.t.sol:71]: This and the USDC test are essentially the same test, just with different input/output assets. You might want to have `_testValidDepositE2E(_inputAsset, _outputAsset, _amount)` instead, and then call that with the different assets to not need the similar tests.

**Review by LHerskind — COMMENTED** [review]
Minor comments, mostly clarity and making the tests a bit more concise.

**Comment by teddav** [comment]
@LHerskind im good :)

**Review by benesjan — APPROVED** [review]

**Pull request merged**
